lib/logstash/plugin_mixins/elasticsearch/common.rb

```diff
@@ -121,6 +79,29 @@ def event_action_tuple(event)
       [action, params, event]
     end
 
+    # Perform some ES options validations and Build the HttpClient.
+    # Note that this methods may sets the @user, @password, @hosts and @client ivars as a side effect.
+    # @param license_checker [#appropriate_license?] An optional license checker that will be used by the Pool class.
+    # @return [HttpClient] the new http client
+    def build_client(license_checker = nil)
```

Comment:
good point. but `metric` is a base plugin getter, IMO `build_client` should not have been calling such a plugin getter  in the first place; passing a metric as parameter would have been cleaner. I am not sure we we should continue down that path?

this process here is a very first step at *trying* to untangle more *generic* code from the more specific plugin code. This metric thing is a perfect example of not properly separating concerns (this is totally blameless, this plugin has a huge history). `build_client` is being used in other contexts than an actual plugin and if this code had been better encapsulated it would have been a lot easier.